src/main/scala/com/microsoft/kusto/spark/datasource/KustoRelation.scala

```diff
@@ -17,34 +17,59 @@ case class KustoRelation(cluster: String,
                          authorityId: String,
                          query: String,
                          isLeanMode: Boolean,
-                         customSchema: Option[String] = None)
-                   (@transient val sparkContext: SQLContext) extends BaseRelation with TableScan with Serializable {
+                         numPartitions: Int,
+                         partitioningColumn: Option[String],
+                         partitioningMode: Option[String],
+                         customSchema: Option[String] = None,
+                         storageAccount: Option[String] = None,
+                         storageContainer: Option[String] = None,
+                         storageAccountSecrete: Option[String] = None,
+                         isStrogageSecreteKeyNotSas: Boolean = true)
+                        (@transient val sparkSession: SparkSession) extends BaseRelation with TableScan with Serializable {
 
   private val normalizedQuery = KustoQueryUtils.normalizeQuery(query)
 
-  override def sqlContext: SQLContext = sparkContext
+  override def sqlContext: SQLContext = sparkSession.sqlContext
 
   override def schema: StructType = {
-    if (customSchema.isDefined)
-    {
+    if (customSchema.isDefined) {
       StructType.fromDDL(customSchema.get)
     }
     else getSchema
   }
 
   override def buildScan(): RDD[Row] = {
-    if (isLeanMode) leanBuildScan() else scaleBuildScan()
+    if (isLeanMode) {
+      KustoReader.leanBuildScan(
+        KustoReadRequest(sparkSession, schema, cluster, database, query, appId, appKey, authorityId)
+      )
+    } else {
+      KustoReader.scaleBuildScan(
+        KustoReadRequest(sparkSession, schema, cluster, database, query, appId, appKey, authorityId),
+        getTransientStorageParameters(storageAccount, storageContainer, storageAccountSecrete, isStrogageSecreteKeyNotSas),
+        KustoPartitionInfo(numPartitions, getPartitioningColumn(partitioningColumn, isLeanMode), getPartitioningMode(partitioningMode))
+      )
+    }
   }
 
-  def leanBuildScan() : RDD[Row] = {
-    val kustoConnectionString = ConnectionStringBuilder.createWithAadApplicationCredentials(s"https://$cluster.kusto.windows.net", appId, appKey, authorityId)
-    val kustoResult = ClientFactory.createClient(kustoConnectionString).execute(database, normalizedQuery)
-    var serializer = KustoResponseDeserializer(kustoResult)
-    sparkContext.createDataFrame(serializer.toRows, serializer.getSchema).rdd
-  }
+  private def getTransientStorageParameters(
```

Comment:
I think it's better to keep as is, since the method returns 'KustoStorageParameters'.
For 'validate' I'd expect it to return 'Unit' (void) or throw exception